### PR TITLE
Add --quiet flag

### PR DESCRIPTION
**Pull request by mollyIV**
Resolves https://github.com/tuist/tuist/issues/6905

### Short description 📝

> Describe here the purpose of your PR.

### How to test the changes locally 🧐

> Include a set of steps for the reviewer to test the changes locally (see [the documentation](https://docs.tuist.io/contributors/get-started) for reference).

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [ ] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### How to test it? 

I didn't find any unit tests I could add to and it's unlikely we can create the new ones for the changes introduced in this pull request. 

I did manual testing: 

* `tuist build` spits out only warnings and errors; I used `ios_app_with_tests` from fixtures 

with `--quiet` flag 👇 

Screenshot 2024-11-08 at 11 30 10

without `--quiet` flag 👇 

Screenshot 2024-11-08 at 11 30 28

* passing `--quiet` and `--verbose` flags should throw an error

Screenshot 2024-11-08 at 11 29 25

* passing `--quiet` and `--json` flags should throw an error 

Screenshot 2024-11-08 at 11 29 47

1) The acceptance tests 

Do you think we could add some acceptance tests? 

2) The docs update 

> In case the PR introduces changes that affect users, the documentation has been updated

I didn't find any place on the docs where we could put the information about `--quiet` flag - any suggestions?

### CODE REVIEW

File: Sources/tuist/TuistApp.swift
Comment (mollyIV) [inline Sources/tuist/TuistApp.swift:16]: > We’d throw an error if you passed both.

Would we like to throw an error at the top level here in `TuistServer`?

File: Sources/tuist/TuistApp.swift
Comment (mollyIV) [inline Sources/tuist/TuistApp.swift:19]: Can `--json` flag be used with `--quiet` or maybe they should be exclusive too? 

If we wanted to allow having both `--json` and `--quiet`, we couldn't add a `quite` case to logger type.

File: Sources/TuistSupport/Logging/Logger.swift
Comment (mollyIV) [inline Sources/TuistSupport/Logging/Logger.swift:62]: I'm open to more elegant implementation suggestions 🙏

File: Sources/TuistSupport/Logging/Logger.swift
Comment (mollyIV) [inline Sources/TuistSupport/Logging/Logger.swift:104]: > And both commands would only print errors and warnings.

File: Sources/TuistSupport/Logging/Logger.swift
Comment (fortmarek) [inline Sources/TuistSupport/Logging/Logger.swift:62]:
```suggestion
            handler = StandardLogHandler(label: label, logLevel: .warning)
```

Shouldn't this work as well?

In the initial issue, I was proposing to only include `error` logs – but I'm not against surfacing warnings as well. Worth asking for some extra opinions on this or find examples of how `--quiet` works in other tools.

File: Sources/tuist/TuistApp.swift
Comment (fortmarek) [reply Sources/tuist/TuistApp.swift:16]: Yes, I think so 👍

File: Sources/tuist/TuistApp.swift
Comment (fortmarek) [reply Sources/tuist/TuistApp.swift:19]: They are exclusive – it doesn't make sense to use `--quiet` with `json`

**Review by fortmarek — COMMENTED** [review]
I'm aligned with the direction here, thanks for taking this on!

File: Sources/tuist/TuistApp.swift
Comment (mollyIV) [reply Sources/tuist/TuistApp.swift:19]: Alright, so I guess we can also throw an error if both `--json` and `--quiet` are passed, right?

File: Sources/tuist/TuistApp.swift
Comment (mollyIV) [reply Sources/tuist/TuistApp.swift:16]: I was playing with `swift` CLI and it appears that when you pass both `--verbose` and `--quiet` flags, the `--verbose` takes precedence. 

```sh
Demo ❯ swift build --quiet --verbose
Building for debugging...
Write auxiliary file ~/Demo/.build/arm64-apple-macosx/debug/swift-version-7BB6C6736DFA3E78.txt
Write auxiliary file ~/Demo/.build/arm64-apple-macosx/debug/Demo.build/sources
/Applications/Xcode-16.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc -module-name Demo -emit-dependencies -emit-module -emit-module-path ... <MORE VERBOSE LOGS>
~/Demo/Sources/Demo/Demo.swift:2:9: warning: initialization of variable 'greetings' was never used; consider replacing with assignment to '_' or removing it
1 | func foo() {
2 |     var greetings = "Hello, World!"
  |         `- warning: initialization of variable 'greetings' was never used; consider replacing with assignment to '_' or removing it
3 | }
4 |
Build complete! (0.41s)
```

Just wanted to make sure that we still would like to throw error instead? 🤔

File: Sources/TuistSupport/Logging/Logger.swift
Comment (mollyIV) [reply Sources/TuistSupport/Logging/Logger.swift:62]: I can confirm that `swift` (SPM) surfaces the warnings: 
Screenshot 2024-11-08 at 11 10 36

File: Sources/TuistSupport/Logging/Logger.swift
Comment (fortmarek) [inline Sources/TuistSupport/Logging/Logger.swift:61]:
```suggestion
            LoggingSystem.bootstrap(quietLogHandler)
```

nit: I think this reads a bit nicer

File: Sources/TuistSupport/Logging/Logger.swift
Comment (fortmarek) [reply Sources/TuistSupport/Logging/Logger.swift:62]: I think this is fine, let's keep the early return 👍

Comment (fortmarek) [reply Sources/TuistSupport/Logging/Logger.swift:62]: Should we also surface the `.notice` and `.success` level? That's what they seem to be doing – I assume `Build complete!` is not a warning.

Let's say for the following `tuist generate` example output:
```
Loading and constructing the graph (Notice level)
It might take a while if the cache is empty (Info level)
Fetching remote binaries. Hold tight... (Info level)
Finished fetching remote binaries (Info level)
Using cache binaries for the following targets: CNIOLinux (Info level)
Generating workspace Tuist.xcworkspace (Notice level)
Generating project TuistApp (Info level)
Generating project Tuist (Info level)
Project generated. (Success level)
Total time taken: 6.503s (Info level)

The following warnings need attention: (Warning level)
 · We detected outdated dependencies. Please run "tuist install" to update them. (Warning level)
```

I'd expect the following output when run with `tuist generate --quiet`:
```
Loading and constructing the graph (Notice level)
Generating workspace Tuist.xcworkspace (Notice level)
Project generated. (Success level)

The following warnings need attention: (Warning level)
 · We detected outdated dependencies. Please run "tuist install" to update them. (Warning level)
```

How does that sound? @tuist/community-core any opinions on this one?

File: Sources/tuist/TuistApp.swift
Comment (fortmarek) [reply Sources/tuist/TuistApp.swift:24]: I'd throw an error personally but don't feel strongly about it

**Comment by mollyIV** [comment]
> Should we also surface the .notice and .success level? That's what they seem to be doing – I assume Build complete! is not a warning.

We could do that, but it'd require either a new logger, or changes in `StandardLogHandler`, because the current implementation handles `.notice`, `.info`, `.debug` and `.trace` levels the same:

https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/Logging/StandardLogHandler.swift#L47-L48

I'd go for a dedicated `struct QuietLogHandler` implementation, 90% copy-paste of a `StandardLogHandler`. 

What would be your recommendation @fortmarek?

**Comment by fortmarek** [comment]
> We could do that, but it'd require either a new logger, or changes in StandardLogHandler, because the current implementation handles .notice, .info, .debug and .trace levels the same:

That's not what the linked code does – that's only for coloring output. `LogHandler`, when initialized with the `.notice` level, should only output the following message types:
- notice
- warning
- error
- critical

So, we shouldn't have to copy-paste the `StandardLogHandler` at all

**Comment by mollyIV** [comment]
> That's not what the linked code does – that's only for coloring output.

Right, sorry, I missed that. I changed a log level of quiet log handler to `.notice`. However, according to my testing, we may be printing out too much now 🤔 

```
Loading and constructing the graph
It might take a while if the cache is empty
Building scheme App
⚠️  --- xcodebuild: Using the first of multiple matching destinations:
AppCore Write Auxiliary File module.modulemap
AppCore Write Auxiliary File AppCore.hmap
AppCore Write Auxiliary File AppCore_const_extract_protocols.json
AppCore Write Auxiliary File AppCore.SwiftFileList
AppCore Write Auxiliary File AppCore.SwiftConstValuesFileList
AppCore Write Auxiliary File AppCore.LinkFileList
AppCore Write Auxiliary File AppCore-OutputFileMap.json
AppCore Write Auxiliary File 7187679823f38a2a940e0043cdf9d637-common-args.resp
AppCore Write Auxiliary File AppCore.DependencyMetadataFileList
AppCore Write Auxiliary File AppCore-project-headers.hmap
AppCore Write Auxiliary File AppCore-own-target-headers.hmap
AppCore Write Auxiliary File AppCore-generated-files.hmap
AppCore Write Auxiliary File AppCore-all-target-headers.hmap
AppCore Write Auxiliary File AppCore-all-non-framework-target-headers.hmap
App Write Auxiliary File App.LinkFileList
App Write Auxiliary File App-OutputFileMap.json
App Write Auxiliary File App.SwiftFileList
App Write Auxiliary File App.SwiftConstValuesFileList
App Write Auxiliary File App_const_extract_protocols.json
App Write Auxiliary File Entitlements-Simulated.plist
App Write Auxiliary File App.hmap
App Write Auxiliary File App.DependencyMetadataFileList
App Write Auxiliary File App-project-headers.hmap
App Write Auxiliary File App-own-target-headers.hmap
App Write Auxiliary File App-generated-files.hmap
App Write Auxiliary File App-all-target-headers.hmap
App Write Auxiliary File App-all-non-framework-target-headers.hmap
App Write Auxiliary File App-DebugDylibPath-normal-arm64.txt
App Write Auxiliary File App-DebugDylibInstallName-normal-arm64.txt
AppCore Copy module.modulemap -> module.modulemap
AppCore Processing AppCore-Info.plist
AppCore Compiling AppCore.swift
AppCore Compiling AppCore.swift
AppCore Copy arm64-apple-ios-simulator.abi.json -> AppCore.abi.json
AppCore Copy arm64-apple-ios-simulator.swiftmodule -> AppCore.swiftmodule
AppCore Copy arm64-apple-ios-simulator.swiftsourceinfo -> AppCore.swiftsourceinfo
AppCore Compiling AppCore_vers.c
App Compiling AppDelegate.swift
App Compiling AppDelegate.swift
Signing AppCore.framework (in target 'AppCore' from project 'App')
AppCore Touching AppCore.framework
App Copy arm64-apple-ios-simulator.swiftdoc -> App.swiftdoc
App Copy arm64-apple-ios-simulator.swiftmodule -> App.swiftmodule
App Copy arm64-apple-ios-simulator.abi.json -> App.abi.json
App Copy arm64-apple-ios-simulator.swiftsourceinfo -> App.swiftsourceinfo
App Copy AppCore.framework -> AppCore.framework
Signing AppCore.framework (in target 'App' from project 'App')
App Processing Info.plist
Signing App.debug.dylib (in target 'App' from project 'App')
Signing __preview.dylib (in target 'App' from project 'App')
Signing App.app (in target 'App' from project 'App')
App Touching App.app
Build Succeeded[0m
The project built successfully[0m[1m[0m
The following warnings need attention:[0m[1m[0m
Tuist/Config.swift is deprecated. Rename Tuist/Config.swift to Tuist.swift at the root.
Program ended with exit code: 0
```

**Comment by fortmarek** [comment]
> Right, sorry, I missed that. I changed a log level of quiet log handler to .notice. However, according to my testing, we may be printing out too much now 🤔

Those logs look all coming from `xcodebuild`/`xcbeautify`. I'd recommend checking out where exactly those logs are coming from and checking the level is correct.

**Comment by mollyIV** [comment]
Hi @fortmarek, 

> Those logs look all coming from xcodebuild/xcbeautify. I'd recommend checking out where exactly those logs are coming from and checking the level is correct.

Indeed, majority of these logs are coming from `xcodebuild` command. Tuist uses `logger.notice` to print them out to the console:

* https://github.com/tuist/tuist/blob/main/Sources/TuistAutomation/XcodeBuild/XcodeBuildController.swift#L312

What we could do now? Few options, I see: 

1) Lower down a log level for the`xcodebuild` output to `debug` - probably a breaking change. 
2) Sticking to a `.warning` level and embrace the fact that tuist would spit out different output in `--quiet` mode than SPM. 
3) Brainstorm @pepicrft idea about introducing `--xcodebuild-log-quiet` to suppress `xcodebuild` output only. He mentioned it [here](https://community.tuist.io/t/enhancing-tuist-output-with-xcbeautify-seeking-customization-options/121/3 ).

WDYT?

**Comment by mollyIV** [comment]
> Why would lowering down the log level be a breaking change? I don't think it will be and I don't see a good reason why xcodebuild output would be printed using the .notice.

Indeed, the default logger used in tuist is a level `.info` - no breaking change here. 

I did one more test with the newest changes, the output with `-quiet` flag looks good 🎉 

Screenshot 2024-11-18 at 10 54 16

File: Sources/tuist/TuistApp.swift
Comment (fortmarek) [inline Sources/tuist/TuistApp.swift:10]: I think this can be called `TuistApp`, I don't know why it was called `TuistServer`, but it's definitely not accurate anymore. Sorry for the fly-by change request, I know it's not directly related to your PR!

File: Sources/tuist/TuistApp.swift
Comment (fortmarek) [inline Sources/tuist/TuistApp.swift:18]: What do you think about defining an array of exclusive options and test for any possible combination? `--json` and `--verbose` should be also exclusive.

**Review by fortmarek — APPROVED** [review]
I'm good with this change 👍 Would wait for one extra review from @tuist/community-core.

**Review by danieleformichelli — APPROVED** [review]

**Comment by fortmarek** [comment]
thanks again for contributing this!

**Pull request merged**
